Aplib.Core.Tests/Desire/GoalStructureTests.cs

```diff
@@ -159,6 +159,44 @@ public void FirstOfGoalStructure_WhenGoalIsUnfinished_ShouldReturnGoal()
         currentGoal.Should().Be(goal);
     }
 
+    [Fact]
+    public void FirstOfGoalStructure_WhenReset_ShouldResetChildren()
+    {
+        // Arrange
+        Mock<IGoalStructure<IBeliefSet>> goalStructure1 = new();
+        Mock<IGoalStructure<IBeliefSet>> goalStructure2 = new();
+
+        Mock<FirstOfGoalStructure<IBeliefSet>> firstOfGoalStructure
+            = new(goalStructure1.Object, goalStructure2.Object) { CallBase = true };
+
+        // Act
+        firstOfGoalStructure.Object.Reset();
+
+        // Assert
+        goalStructure1.Verify(x => x.Reset(), Times.Once);
+        goalStructure2.Verify(x => x.Reset(), Times.Once);
+    }
+
+    [Fact]
+    public void FirstOfGoalStructure_WhenReset_ShouldBeUnfinished()
+    {
+        // Arrange
+        Mock<IGoalStructure<IBeliefSet>> goalStructure1 = new();
+        goalStructure1.SetupGet(g => g.Status).Returns(CompletionStatus.Success);
+        Mock<IGoalStructure<IBeliefSet>> goalStructure2 = new();
+        goalStructure2.SetupGet(g => g.Status).Returns(CompletionStatus.Success);
+
+        IBeliefSet beliefSet = Mock.Of<IBeliefSet>();
+        FirstOfGoalStructure<IBeliefSet> firstOfGoalStructure = new(goalStructure1.Object, goalStructure2.Object);
+
+        // Act
+        firstOfGoalStructure.UpdateStatus(beliefSet);
+        firstOfGoalStructure.Reset();
+
+        // Assert
+        firstOfGoalStructure.Status.Should().Be(CompletionStatus.Unfinished);
+    }
```

Comment:
I think this test should check that the status was not `Unfinished` in-between. Currently, this test also passes if you comment out the act section.

Reply:
Maybe like this:

```csharp
    [Fact]
    public void FirstOfGoalStructure_WhenReset_ShouldBeUnfinished()
    {
        // Arrange
        Mock<IGoalStructure<IBeliefSet>> goalStructure1 = new();
        goalStructure1.SetupGet(g => g.Status).Returns(CompletionStatus.Success);
        Mock<IGoalStructure<IBeliefSet>> goalStructure2 = new();
        goalStructure2.SetupGet(g => g.Status).Returns(CompletionStatus.Success);

        IBeliefSet beliefSet = Mock.Of<IBeliefSet>();
        FirstOfGoalStructure<IBeliefSet> firstOfGoalStructure = new(goalStructure1.Object, goalStructure2.Object);

        // Act
        firstOfGoalStructure.UpdateStatus(beliefSet);
        CompletionStatus statusAfterUpdate = firstOfGoalStructure.Status;
        firstOfGoalStructure.Reset();

        // Assert
        statusAfterUpdate.Should().Be(CompletionStatus.Success);
        firstOfGoalStructure.Status.Should().Be(CompletionStatus.Unfinished);
    }
```

Reply:
Or like this:
```csharp
    [Fact]
    public void FirstOfGoalStructure_WhenReset_ShouldBeUnfinished()
    {
        // Arrange
        Mock<IGoalStructure<IBeliefSet>> goalStructure1 = new();
        goalStructure1.SetupGet(g => g.Status).Returns(CompletionStatus.Success);
        Mock<IGoalStructure<IBeliefSet>> goalStructure2 = new();
        goalStructure2.SetupGet(g => g.Status).Returns(CompletionStatus.Success);

        IBeliefSet beliefSet = Mock.Of<IBeliefSet>();
        FirstOfGoalStructure<IBeliefSet> firstOfGoalStructure = new(goalStructure1.Object, goalStructure2.Object);

        // Pre-conditions
        firstOfGoalStructure.UpdateStatus(beliefSet);
        firstOfGoalStructure.Status.Should().Be(CompletionStatus.Success);

        // Act
        firstOfGoalStructure.Reset();

        // Post-conditions
        firstOfGoalStructure.Status.Should().Be(CompletionStatus.Unfinished);
    }
```

Reply:
I just read a little bit about this, and it seems this is a different testing pattern than is usual in unit testing. See [this StackOverflow question](https://stackoverflow.com/questions/1021007/should-it-be-arrange-assert-act-assert). I don't think we have done this way of unit testing before. It also is not what is tested; if we want to assert that the setup is correct, we should write a separate test for that (I think the behavior of the Arrange code in these tests is already properly tested in our test suite).

I do agree that UpdateStatus should not be in Act, so I moved that up into Arrange.